File: 210.course-schedule-ii.go
/*
 * @lc app=leetcode id=210 lang=golang
 *
 * [210] Course Schedule II
 */

// @lc code=start
func findOrder(numCourses int, prerequisites [][]int) []int {
	in := make([]int, numCourses)
	frees := make([][]int, numCourses)
	next := make([]int, 0, numCourses)

	for _, courses := range prerequisites {
		in[courses[0]]++
		frees[courses[1]] = append(frees[courses[1]], courses[0])
	}

	for c, i := range in {
		if i == 0 {
			next = append(next, c)
		}
	}

	for i := 0; i != len(next); i++ {
		c := next[i]

		for _, v := range frees[c] {
			in[v]--
			if in[v] == 0 {
				next = append(next, v)
			}
		}
	}

	if len(next) == numCourses {
		return next
	}

	return []int{}
}

// @lc code=end

File: 210.course-schedule-ii.cpp
/*
 * @lc app=leetcode id=210 lang=cpp
 *
 * [210] Course Schedule II
 */

// @lc code=start
#include <vector>

class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        int n = numCourses;

        vector<int> in(n, 0);
        vector<vector<int>> frees(n);
        vector<int> next;

        for (auto& prereq : prerequisites) {
            in[prereq[0]]++;
            frees[prereq[1]].push_back(prereq[0]);
        }

        for (int c = 0; c < in.size(); ++c) {
          if (in[c] == 0) {
            next.push_back(c);
          }
        }

        int i = 0;
        while (i < next.size()) {
          int c = next[i];

          for (int v : frees[c]) {
            in[v]--;

            if (in[v] == 0) {
              next.push_back(v);
            }
          }

          ++i;
        }

        if(next.size() == n) return next;

        return {};
    }
};
// @lc code=end

